pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go

```diff
@@ -283,9 +283,16 @@ func computeKindTypesForShoots(
 // Seed configuration
 func computeKindTypesForSeed(
 	seed *gardencorev1beta1.Seed,
+	numberOfShoots int,
+	numberOfBuckets int,
 ) sets.String {
 	var wantedKindTypeCombinations = sets.NewString()
 
+	seedInDeletionAndEmpty := seed.ObjectMeta.DeletionTimestamp != nil && numberOfBuckets == 0 && numberOfShoots == 0
```

Comment:
Below logic only evaluates the `.spec.dns` field today, so why is the `numberOfBuckets` playing a role here?

Reply:
We want to keep access to logging and monitoring until everything is cleaned up. So if deletion of Shoots or Backupbucket fails you can still debug it.

Reply:
Hm, not sure how this answer is related, but isn't it conceptually wrong to check for buckets/shoots here? I mean, this is maybe a "good?" approximation, but not the "correct" solution, right? The `DNSProvider/DNSEntry` resources in the `Seed` could still be present even if there are no `Shoot`s/`BackupBucket`s anymore (maybe the DNS secret is wrong and they cannot be deleted - then an operator needs to change it first).

The `Seed` deletion is a bit special, I agree. I would propose to rather adapt https://github.com/gardener/gardener/blob/0a5df55ecf3a3824662450a7c8d6498d55605215/pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go#L319-L371. I think we can skip computing the names of the required `ControllerRegistration`s if the `Seed` has a `deletionTimestamp` and the `Required` condition of the backing `ControllerInstallation` is `False`.

As concrete implementation proposal:

1. Move https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go#L359-L367 to https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go#L349
2. Before https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go#L366, store the name of the `ControllerRegistration` in a new set `requiredControllerRegistrationNames`. At this point we know that there is at least one `ControllerInstallation` with `Required=True` referencing the `ControllerRegistration`.
3. In https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go#L355, if the `Seed`'s `deletionTimestamp` is set then iterate over `names` and only insert the name to `wantedControllerRegistrationNames` in case it exists in `requiredControllerRegistrationNames`.

Reply:
Done

Reply:
As I addressed you remarks please take another look @rfranzke